File: LeetCode_Java/hash/GroupAnagrams.java
package hash;

import java.util.*;

public class GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] strs) {

        Map<String, List<String>> map = new HashMap<>();

        for (String string : strs) {
            char[] letters = string.toCharArray();
            Arrays.sort(letters);
            String sortedString = new String(letters);

            map.computeIfAbsent(sortedString, k -> new ArrayList<>()).add(string);
        }

        return new ArrayList<>(map.values());
    }

    public static void main(String[] args) {
        GroupAnagrams solution = new GroupAnagrams();

        String[] test1 = {"eat", "tea", "tan", "ate", "nat", "bat"};
        List<List<String>> result1 = solution.groupAnagrams(test1);
        System.out.println(result1);  // Output: [[bat], [nat, tan], [ate, eat, tea]]

        String[] test2 = {""};
        List<List<String>> result2 = solution.groupAnagrams(test2);
        System.out.println(result2);  // Output: [[""]]

        String[] test3 = {"a"};
        List<List<String>> result3 = solution.groupAnagrams(test3);
        System.out.println(result3);  // Output: [["a"]]
    }
}

File: LeetCode_Python/hash/GroupAnagrams.py
import collections


class GroupAnagrams:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        res = collections.defaultdict(list)

        for s in strs:
            key = tuple(sorted(s))
            res[key].append(s)

        return list(res.values())


if __name__ == "__main__":
    solution = GroupAnagrams()

    print('Expected Output: [["bat"],["nat","tan"],["ate","eat","tea"]]')
    print(
        "Actual Output:",
        solution.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]),
    )

    print('Expected Output: [[""]]')
    print("Actual Output:", solution.groupAnagrams([""]))

    print('Expected Output: [["a"]]')
    print("Actual Output:", solution.groupAnagrams(["a"]))
